Mathlib/Data/Nondet/Basic.lean

```diff
@@ -0,0 +1,176 @@
+/-
+Copyright (c) 2023 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import Std.Tactic.Lint
+import Std.Data.MLList.Basic
+import Mathlib.Data.MLList.Basic
+
+/-!
+# A nondeterminism monad.
+
+We represent nondeterministic values in a type `α` as `MLList m (σ × α)`,
+i.e. as a monadic lazy list of possible values, each equipped with the backtrackable state
+required to run further computations in the ambient monad.
+
+We define the type synonym `Nondet m α` for this, as we need to equip it with a different
+`Monad` instance (that backtracks state appropriately in `bind`) than the default provided
+for `MLList`.
+
+Similarly, various monadic functions need to be redefined (e.g. `mapM` and `filterMapM`),
+and to avoid confusion resulting from using the `MLList` API,
+we eventually mark `Nondet` as irreducible.
+-/
+
+set_option autoImplicit true
+
+open Lean
+
+/--
+`Nondet m α` is variation on `MLList m α` suitable for use with backtrackable monads `m`.
+
+We think of `Nondet m α` as a nondeterministic value in `α`,
+with the possible alternatives stored in a monadic lazy list.
+
+Along with each `a : α` we store the backtrackable state, and ensure that monadic operations
+on alternatives run with the appropriate state.
+-/
+@[nolint unusedArguments]
+def Nondet (m : Type → Type) [MonadBacktrack σ m] (α : Type) : Type := MLList m (σ × α)
+
+namespace Nondet
+
+variable {m : Type → Type} [Monad m] [MonadBacktrack σ m]
+
+instance : Inhabited (Nondet m α) := ⟨MLList.nil⟩
+
+/--
+Bind a nondeterministic function over a nondeterministic value,
+ensuring the function is run with the relevant backtrackable state at each value.
+-/
+partial def bind (L : Nondet m α) (f : α → Nondet m β) : Nondet m β :=
+  L.casesM
+    (fun _ => pure .nil)
+    (fun ⟨s, x⟩ xs => do match ← (do restoreState s; MLList.uncons (f x)) with
+    | none => return bind xs f
+    | some (y, ys) => return .cons y (ys.append (fun _ => bind xs f)))
+
+/-- `Nondet m` is a monad. -/
+instance : Monad (Nondet m) where
+  pure a := .singletonM do return (← saveState, a)
+  bind := bind
+
+/-- `Nondet m` is an alternative monad. -/
+instance : Alternative (Nondet m) where
+  failure := MLList.nil
+  orElse := MLList.append
+
+/-- Convert any value in the monad to the singleton nondeterministic value. -/
+def monadLift (x : m α) : Nondet m α :=
+  .singletonM do
+    let a ← x
+    return (← saveState, a)
+
+instance : MonadLift m (Nondet m) where
+  monadLift := monadLift
+
+/--
+Lift a list of monadic values to a nondeterministic value.
+We ensure that each monadic value is evaluated with the same backtrackable state.
+-/
+def ofListM [Alternative m] (L : List (m α)) : Nondet m α :=
+  MLList.squash fun _ => do
+    let s ← saveState
+    return MLList.ofList L |>.mapM fun x => do
+      restoreState s
+      let a ← x
```

Comment:
What happens if `x` fails here? Is the expectation that the state represents the point of failure, or that it backtracks to the state before the failure?

Reply:
This is incorrect, I want it to return .nil here.

Reply:
@eric-wieser, I'm not going to answer specifically here, as prompted by the (excellent) suggestion to use a one-field structure I've completely rewritten the file.

But failures are just meant to propagate. If you want to backtrack on failures then you should wrap the failures in `try?` and then use `filterMapM` to discard the `none` values.

I think this is "the Lean 4 way".